### PR TITLE
Large collateral ir ema test

### PR DESCRIPTION
**Pull request by mattcushman**
# Added test for Sherlock issue 100 - pledging large amounts of collateral
## High level
* Added `testPoolLargeCollateralPostedTargetUtilization`
* Also fixed a comment

# Description of bug or vulnerability and solution
* Sherlock issue 100 was not tested
* Sherlock issue 100 observes that the pool was vulnerable to an attack where a borrower could pledge huge amounts of collateral to the pool and draw no or little debt, driving down the target utilization.  This was fixed by weighting loans by debt in the TU calculation and introducing new EMAs to smooth out the interest rate calculation.  This tests the scenario of a borrower pledging massive collateral to make a very underlevered loan as in this attack vector.
# Contract size
## Pre Change
no impact (test only)
## Post Change
no impact (test only)

# Gas usage
## Pre Change
no impact (test only)
## Post Change
no impact (test only)

### CODE REVIEW

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 1]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                actualUtilization:    0,
+                targetUtilization:    1e18,
+                minDebtAmount:        0,
+                loans:                0,
+                maxBorrower:          address(0),
+                interestRate:         0.05 * 1e18,
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:708]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   0,
            lupColEma: 0
        });
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 2]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+
+        _assertPool(
+            PoolParams({
+                htp:                  200.192307692307692400 * 1e18,
+                lup:                  327.188250324085203338 * 1e18,
+                poolSize:             40_000 * 1e18,
+                pledgedCollateral:    50 * 1e18,
+                encumberedCollateral: 30.592832642066761971 * 1e18,
+                poolDebt:             10009.615384615384620000 * 1e18,
+                actualUtilization:    0,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:736]: w/o 4 EMAs `actualUtilization:    0.250240384615384616 * 1e18,`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 3]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                targetUtilization:    1e18,
+                minDebtAmount:        1000.961538461538462000 * 1e18,
+                loans:                1,
+                maxBorrower:          address(_borrower),
+                interestRate:         0.05 * 1e18,
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:745]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   0,
            lupColEma: 0
        });
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 4]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+
+        _assertPool(
+            PoolParams({
+                htp:                  200.192307692307692400 * 1e18,
+                lup:                  327.188250324085203338 * 1e18,
+                poolSize:             40_000 * 1e18,
+                pledgedCollateral:    100 * 1e18,
+                encumberedCollateral: 58.126382019926847745 * 1e18,
+                poolDebt:             19_018.269230769230778000 * 1e18,
+                actualUtilization:    0,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:773]: w/o 4 EMAs `actualUtilization:    0.475456730769230769 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 5]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                targetUtilization:    1e18,
+                minDebtAmount:        950.913461538461538900 * 1e18,
+                loans:                2,
+                maxBorrower:          address(_borrower),
+                interestRate:         0.05 * 1e18,
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:782]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   0,
            lupColEma: 0
        });
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 6]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+            amount:     10 * 1e18,
+            indexLimit: 3_010,
+            newLup:     327.188250324085203338 * 1e18
+        });
+
+        _assertPool(
+            PoolParams({
+                htp:                  200.941998518054562754 * 1e18,
+                lup:                  327.188250324085203338 * 1e18,
+                poolSize:             40_022.159734498291800000 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:803]: w/o 4 EMAs `poolSize:             40_022.916501223632360000 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 7]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                pledgedCollateral:    100 * 1e18,
+                encumberedCollateral: 58.236654596124865142 * 1e18,
+                poolDebt:             19_054.349122034189453676 * 1e18,
+                actualUtilization:    0.476358955196505217 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:807]: w/o 4 EMAs
``` 
                actualUtilization:    0.476085972431610156 * 1e18,
                targetUtilization:    0.582366545961248651 * 1e18,
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 8]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                targetUtilization:    0.584010402015984926 * 1e18,
+                minDebtAmount:        952.717456101709472684 * 1e18,
+                loans:                2,
+                maxBorrower:          address(_borrower),
+                interestRate:         0.05 * 1e18,
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:816]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   1_796.374215061630403555 * 1e18,
            lupColEma: 3_084.610933645840358918 * 1e18
        });
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 9]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+        });
+        _borrow({
+            from:       _borrower2,
+            amount:     12 * 1e18,
+            indexLimit: 3_010,
+            newLup:     327.188250324085203338 * 1e18
+        });
+        _assertPool(
+            PoolParams({
+                htp:                  201.493279375818240993 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:839]: w/o 4 EMAs `htp:                  200.941998518054562754 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 10]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                lup:                  327.188250324085203338 * 1e18,
+                poolSize:             40_045.121523671487120000 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:841]: w/o 4 EMAs `poolSize:             40_022.916501223632360000 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 11]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                pledgedCollateral:    50_100 * 1e18,
+                encumberedCollateral: 58.353196900686720280 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:843]: w/o 4 EMAs
```
                encumberedCollateral: 58.273369521846658463 * 1e18,
                poolDebt:             19_066.361814341881761373 * 1e18,
                actualUtilization:    0.476386117782270072 * 1e18,
                targetUtilization:    0.164922801681392437 * 1e18,
                minDebtAmount:        953.318090717094088069 * 1e18,
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 12]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                poolDebt:             19_092.480394752519489737 * 1e18,
+                actualUtilization:    0.476094974846641824 * 1e18,
+                targetUtilization:    0.584010402015984926 * 1e18,
+                minDebtAmount:        954.624019737625974487 * 1e18,
+                loans:                2,
+                maxBorrower:          address(_borrower),
+                interestRate:         0.05 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:850]: w/o 4 EMAs
```
                interestRate:         0.055 * 1e18,
                interestRateUpdate:   _startTime + 10 days
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 13]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:854]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   3_423.392851516449417444 * 1e18,
            lupColEma: 20_757.547268266524846613 * 1e18
        });
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 14]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+        // borrower 3 touches the pool again to force an interest rate update
+        _borrow({
+            from:       _borrower2,
+            amount:     3.14 * 1e18,
+            indexLimit: 3_010,
+            newLup:     327.188250324085203338 * 1e18
+        });
+        _assertPool(
+            PoolParams({
+                htp:                  201.548490576836267720 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:871]: w/o 4 EMAs `htp:                  201.002565506695771731 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 15]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                lup:                  327.188250324085203338 * 1e18,
+                poolSize:             40_047.420826509653040000 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:873]: w/o 4 EMAs `poolSize:             40_025.442356017355120000 * 1e18`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 16]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                pledgedCollateral:    50_100 * 1e18,
+                encumberedCollateral: 58.370797186283932430 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:875]: w/o 4 EMAs
```
                encumberedCollateral: 58.291758174381564242 * 1e18,
                poolDebt:             19_072.378365390595137386 * 1e18,
                actualUtilization:    0.476506373015095162 * 1e18,
                targetUtilization:    0.133184970951579086 * 1e18,
                minDebtAmount:        953.618918269529756869 * 1e18,
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 17]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                poolDebt:             19_098.239001402275530018 * 1e18,
+                actualUtilization:    0.476604459561723991 * 1e18,
+                targetUtilization:    0.584213148132606714 * 1e18, // big col. deposit barely affects
+                minDebtAmount:        954.911950070113776501 * 1e18,
+                loans:                2,
+                maxBorrower:          address(_borrower),
+                interestRate:         0.05 * 1e18,
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:882]: w/o 4 EMAs
```
                interestRate:         0.0605 * 1e18,
                interestRateUpdate:   _startTime + 10 days + 1 days
```

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol [hunk 18]
```diff
@@ -674,6 +674,222 @@ contract ERC20PoolInterestRateTestAndEMAs is ERC20HelperContract {
...
+                interestRateUpdate:   _startTime
+            })
+        );
+        _assertEMAs({
```
Comment (grandizzy) [inline tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:886]: w/o 4 EMAs
```
        _assertEMAs({
            debtEma:   4_898.721863751709369557 * 1e18,
            lupColEma: 36_781.341233559269369810 * 1e18
        });
```

**Review by grandizzy — COMMENTED** [review]
thanks, I think number looks good, I left comment with numbers from same test performed against old code base with 2 EMAs where it can be seen interest rate increase and `targetUtilization` dropping from `0.584010402015984926` to `0.133184970951579086`

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol (hunk 11 above)
Comment (mattcushman) [reply tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:843]: Thanks -- here is where you can start to see the effect of the t0D2toC accumulator affecting the TU

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol (hunk 16 above)
Comment (mattcushman) [reply tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:875]: note the TU difference

File: tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol (hunk 18 above)
Comment (mattcushman) [reply tests/forge/ERC20Pool/ERC20PoolInterestRateAndEMAs.t.sol:886]: Also note how the previous code the lupColEma jumps

**Review by grandizzy — APPROVED** [review]

**Review by EdNoepel — APPROVED** [review]
Thank you for saving me the effort in writing this myself.  See also: https://github.com/ajna-finance/contracts/pull/678

**Comment by grandizzy** [comment]
> Thank you for saving me the effort in writing this myself. See also: #678

credits to @mattcushman for creating the test, I just run it on old code base and made comparisons :)

**Pull request merged**